src/Chapter3.hs

```diff
@@ -1006,6 +1161,79 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data MonsterActions = MAttack | Run
+    deriving (Show, Eq, Enum, Bounded)
+
+data Monster' = M 
+    { mHealth  :: Health
+    , mAttack  :: Attack
+    , mActions :: MonsterActions
+    } deriving Show
+
+data KnightActions 
+    = KAttack
+    | Heal
+    | SpellCast
+    deriving (Show, Eq, Enum, Bounded)
+  
+data Knight' = K
+    { kHealth  :: Health
+    , kAttack  :: Attack
+    , kDefense :: Defense
+    , kActions :: KnightActions
+    } deriving Show
+
+cycleMActions :: MonsterActions -> MonsterActions
+cycleMActions a
+    | a == maxBound   = minBound
+    | otherwise       = succ a
+
+cycleKActions :: KnightActions -> KnightActions
+cycleKActions a
+    | a == maxBound   = minBound
+    | otherwise       = succ a
+
+getMonsterDamage :: Health -> Attack -> Health
+getMonsterDamage (Health h) (Attack a) = Health (h - a)
+
+getKnightDamange :: Health -> Defense -> Attack -> Health
+getKnightDamange (Health h) (Defense d) (Attack a) = Health ((h + d) - a)
+
+skipKAttack :: KnightActions -> Attack -> Attack
+skipKAttack KAttack a = a
+skipKAttack _ _ = Attack 0
+    
+skipMAttack :: MonsterActions -> Attack -> Attack
+skipMAttack MAttack a = a
+skipMAttack _ _ = Attack 0
+
+class Battle a b where {
+    battle :: a -> b -> String
+}
```

Comment:
I didn't know you can use `{}` in the typeclass definition :laughing: Though, usually, they are not needed
```suggestion
class Battle a b where
    battle :: a -> b -> String
```